src/twisted/logger/test/_assertionhelpers.py

```diff
@@ -0,0 +1,244 @@
+# -*- test-case-name: twisted.logger.test.test_assertionhelpers -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+General assertion helpers for testing log events.
+"""
+
+from contextlib import contextmanager
+
+from twisted.logger import globalLogPublisher
+from twisted.logger._legacy import _ensureCompatWithLegacyObservers
+
+
+
+@contextmanager
+def capturingLogEventsFrom(logger):
+    """
+    Context manager to accumulate log events emitted by a L{Logger}
+    instance by passing list.append as an observer.
+
+    @param logger: L{Logger} object that the event was emitted to.
+    @type logger: L{Logger}
+    """
+    events = []
+    logger.observer = events.append
```

Comment:
`Logger` implements [the descriptor protocol](https://docs.python.org/2.7/howto/descriptor.html) so that each access to either enclosing class or instance results in [a unique logger instance](https://github.com/twisted/twisted/blob/trunk/src/twisted/logger/_logger.py#L88-L92).  That means this assignment won't have any effect on subsequently created loggers.  You can verify this in a Python shell:

```
>>> from twisted.logger.test.test_assertionhelpers import ThingThatLogs
>>> ThingThatLogs._log is ThingThatLogs._log
False
>>> thing = ThingThatLogs()
>>> thing._log is thing._log
False
```

The next if statement appears to work around this by reassigning the attribute on the thing to which the `Logger`'s bound.  Something like this:

```
>>> thing._log = thing._log
>>> thing._log is thing._log
True
```

There are several problems with this approach.

1) It only works if the instance (`source`) has named its `Logger` `_log`.  Consider the following:
```python
from twisted.logger import Logger
from twisted.logger.test._assertionhelpers import capturingLogEventsFrom


class SomeClass(object):
    log = Logger()

    def logSomething(self):
        self.log.info("a message")


instance = SomeClass()
with capturingLogEventsFrom(instance.log) as events:
    instance.logSomething()

print(events)  # []
```

In fact, it will fail loudly if the class has defined slots:

```python
class HasSlots(object):
    __slots__ = ['attribute']
    log = Logger()

    def __init__(self, attribute):
        self.attribute = attribute

    def logAttribute(self):
        self.log.info("attribute: {attribute!}", attribute=self.attribute)


instance = HasSlots("value")
with capturingLogEventsFrom(instance.log) as events:  # AttributeError: 'HasSlots' object has no attribute '_log'
    instance.logAttribute()
```

2. The instance has been permanently modified.  Even after the context manager's block completes, the observer will still be the `append` method of that particular `events` list!  That particular issue can be solved by saving and restoring the original observer, like so:

```python
@contextmanager
def capturingLogEventsFrom(logger):
    ...
    events = []
    originalObserver = logger.observer
    logger.observer = events.append
    if logger.source is not None:
        logger.source._log = logger
    try:
        yield events
    finally:
        logger.observer = originalObserver
```

This is still problematic, though, because now `logger.source._log` is a single static instance instead of one that possesses `Logger`'s binding behavior 

1) is worse than 2) and has a simpler solution: change the signature of `capturingLogEventsFrom` to accept the instance and name of logger on that instance.  Like so:

```python
def capturingLogEventsFrom(something, loggerName):
    logger = getattr(something, loggerName)
    ...
```

But there's a solution to 2) here as well, though it's a dirty hack.  We can circumvent the descriptor protocol by accessing the `Logger` via the instance, or its class', `__dict__`:
```python
@contextmanager
def capturingLogEventsFrom(something, loggerName):
    logger = something.__dict__.get(
        loggerName,
        something.__class__.__dict__[loggerName])
    events = []
    logger.observer = events.append
    if logger.source is not None:
        logger.source._log = logger

    yield events
```

I'd prefer if we modified `Logger` to track is originating instance so that we could set the observer there.  I'll let other reviewers comment on this.